src/typechecker/FStar.TypeChecker.Tc.fst

```diff
@@ -96,10 +96,24 @@ let tc_assume (env:env) (ts:tscheme) (r:Range.range) :tscheme =
   //AR: this might seem same as tc_declare_typ but come prop, this will change
   tc_type_common env ts (U.type_u () |> fst) r
 
+let tc_decl_attributes env se =
+  // [Substitute] (defined in Pervasives), is added as attribute
+  // by TcInductive when a type has no projector, and this
+  // happens for some types (see
+  // TcInductive.early_prims_inductives) that are defined before
+  // Substitute even exists.
+  // Thus the hack below, that typecheck attribtues only when
+  // they're not literally [Substitute]. TODO: fix this.
```

Comment:
Perhaps you can guard this with `FStar.TypeChecker.Env.too_early_in_prims ()`? Or at least checking to see if `Substitute` is already present in the environment.

Reply:
Didn't know about `too_early_in_prims`, thanks.
I just tried using `too_early_in_prims`, but it seems like the inductive `dtuple2` in `prims` is also tagged with attribute `Substitute`, but I don't understand why: it is not in the `early_prims_inductives` list, and it is not tagged with attribute `"no_auto_projectors"`.